Add a GTK4 backend.

QuLogic opened the pull request:
## PR Summary

This is currently sufficient to open a window, do the usual things, like handle key presses, or zoom/pan, but there are still some kinks to work out. The zoom/pan buttons do not stay pressed, the canvas does not resize, and there are probably several other smaller things that are broken.

It is also in a separate file, but several of the changes can be cross-ported to the GTK3 backend, which I will do separately. When that is done, I can likely remove a lot of duplication, and possibly put it all together like Qt5/Qt6 is aiming to do.

## PR Checklist

- [ ] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [ ] New features are documented, with examples if plot related.
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [x] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

h-vetinari commented on lib/matplotlib/backends/backend_gtk4.py:
```diff
@@ -0,0 +1,808 @@
+_log = logging.getLogger(__name__)
+
+backend_version = "%s.%s.%s" % (
+    Gtk.get_major_version(), Gtk.get_micro_version(), Gtk.get_minor_version())
```
looks like minor/micro are switched...

QuLogic replied:
Ah looks like this has been wrong in the GTK3 backend too, since it was introduced.

QuLogic commented:
I've now fixed resizing, which was easier to fix than I first thought, so it's just the toolbar button setup that needs fixing. Then there's lots of shared code between 2 and 3 that could be refactored out.

QuLogic commented:
OK, I've fixed the toolbar, as well as tool manager, a focus issue, and this seems to be working in general. I haven't done the refactor or updated CI yet, but it is in a testable state.

QuLogic commented:
OK, I noticed some more bugs and have fixed them now. The tool manager copy button works now, and save figure uses the native dialog.

I moved common code into a separate file. I'm debating whether the canvas and manager should be refactored together. The event handling is quite a bit different, so this may not save much. Really only the draw event handling (with a rubberband) is common on the canvas.

IPython does not have an entry for gtk4, so does not set up event loop integration. I don't know if there's anything we can do on our side to get that working.

anntzer commented on lib/matplotlib/backends/_backend_gtk.py:
```diff
@@ -0,0 +1,183 @@
+        class PseudoToolbar:
+            canvas = self.figure.canvas
+
+        return _NavigationToolbar2GTK.save_figure(PseudoToolbar())
```
This doesn't work (whether on gtk3 or gtk4), as save_figure now resolves to the backend_bases' save_figure (which raises NotImplementedError).

QuLogic replied:
Oops, when I found the broken stuff, I tested on the original change, and forgot to do it after refactoring.

anntzer commented on lib/matplotlib/backends/backend_gtk4.py:
```diff
@@ -0,0 +1,683 @@
+            # Some versions of gtk throw a glib.GError but not all, so I am not
+            # sure how to catch it.  I am unhappy doing a blanket catch here,
+            # but am not sure what a better way is - JDH
+            _log.info('Could not load matplotlib icon: %s', sys.exc_info()[1])
```
This log is emitted, because `'Window' object has no attribute 'set_icon_from_file'` (perhaps we should indeed get rid of the blanket catch, as suggested above, to avoid hiding this).

QuLogic replied:
Right, we need to do something completely different on GTK4, for which I will open a separate PR as it needs to be done in many backends.

anntzer replied:
Just to be clear, this will stay unimplemented in this PR?

QuLogic replied:
Yes, it needs some stuff that will need to go in Qt also, and/or a restructuring of image directories.

QuLogic commented:
Fixed, and I also added docs and example updates. Since we don't need to support GTK 3.0.0, I used the latest recommended style for the embedding examples.

anntzer reviewed (dismissed):
Modulo rebase, CI, and clarifying https://github.com/matplotlib/matplotlib/pull/20321#discussion_r684411273.

anntzer commented on lib/matplotlib/backends/backend_gtk4.py:
```diff
@@ -0,0 +1,683 @@
+        # TODO: Only update the rubberband area.
+        self.queue_draw()
+
+    def draw_func(self, drawing_area, ctx, width, height):
```
Keep this private?  (this makes is slightly clearer that 3rd party renderers (i.e., mplcairo.gtk) can keep on overriding on_draw_event as in gtk3, not something else)

anntzer commented:
Actually I found another bug: with this PR, on gtk3 (not gtk4), if you press `o` to toggle the zoom tool, you get an infinite recursion error (repeatedly going between `_update_buttons_checked()` and `zoom()`).

anntzer commented:
Actually I found another bug: with this PR, on gtk3 (not gtk4), if you press `o` to toggle the zoom tool, you get an infinite recursion error (repeatedly going between `_update_buttons_checked()` and `zoom()`).

anntzer commented:
Actually I found another bug: with this PR, on gtk3 (not gtk4), if you press `o` to toggle the zoom tool, you get an infinite recursion error (repeatedly going between `_update_buttons_checked()` and `zoom()`).

anntzer commented:
Actually I found another bug: with this PR, on gtk3 (not gtk4), if you press `o` to toggle the zoom tool, you get an infinite recursion error (repeatedly going between `_update_buttons_checked()` and `zoom()`).

QuLogic commented:
Ah, I thought the block wasn't working in GTK4, but it was really just a stylistic issue. Adding the `flat` class made the changes correctly visible, so we can go back to the same blocking implementation that worked on GTK3.

anntzer commented:
I can't say I checked everything, but this seems good enough to go for now and we can always fix anything else we find later.

QuLogic commented:
@tacaswell The last commit fixes <kbd>Ctrl+C</kbd>, but it needs the IPython input hooks to work completely as expected ipython/ipython#13101

tacaswell commented:
I @QuLogic can self-merge this after a rebase and CI green.
